### ISSUE TITLE
Support and validate client-defined losses (forward_backward_custom) across HF and FSDP

### ISSUE THREAD

**Issue by ekzhu** [agentscope-ai/TuFT#146]
## Problem

TuFT supports a fixed set of server-side loss names, but it does not currently document or test the Tinker SDK's client-defined `forward_backward_custom` workflow as a supported TuFT capability. This blocks portable examples that need objectives beyond the built-in CE/RL losses.

The Open Character Training recipe is a concrete case. Its DPO stage combines:

```text
DPO(beta=0.1)
+ 0.1 * chosen response-token mean NLL
+ 0.001 * chosen/rejected mean squared policy/reference log-ratio
```

Its SFT stage uses the batch mean of per-example response-token mean NLL. The DPO objective also couples chosen/rejected rows across the logical batch.

To run this today, we had to patch a TuFT deployment with named experiment-specific losses. Because the installed client SDK restricts the public loss-name type to its built-in names, the deployment then had to dispatch those losses through `dro` and `cross_entropy` using a sentinel `loss_fn_config` value. That workaround is not appropriate for an upstream example: it requires editing TuFT source and can silently diverge across HF and FSDP.

The SDK already exposes a safer general mechanism: `forward_backward_custom`. It performs a forward pass to obtain per-token target log probabilities, differentiates a client-defined PyTorch scalar loss with respect to those log probabilities, and sends the resulting token weights through the standard backward path. TuFT should explicitly support and validate this workflow rather than execute arbitrary client Python on the server.

## Proposed scope

1. Treat `TrainingClient.forward_backward_custom(..., loss_type_input="logprobs")` as a supported compatibility path for both HF and FSDP backends.
2. Add end-to-end tests using the pinned Tinker SDK version. Keep the tests when the SDK 0.25 protocol work in #145 lands.
3. Document the two-pass semantics, supported input space, performance cost, error behavior, and the requirement that the model state remain unchanged between the forward and backward halves.
4. Verify that arbitrary datum `loss_fn_inputs` remain aligned with returned per-datum log probabilities and token-weight gradients. #137 supplies the generic input preservation needed here.
5. Keep the server allowlisted: custom Python remains client-side. An extensible server-side named-loss/plugin API can be a separate feature if maintainers want it.

## Suggested tests

- **HF smoke:** a simple custom weighted log-probability loss completes `forward_backward_custom`, `optim_step`, and changes the adapter weights.
- **FSDP smoke:** the same callback completes with two FSDP ranks and produces finite gradients/metrics.
- **Built-in parity:** one update of a custom callback analytically equivalent to weighted cross-entropy matches the built-in loss within tolerance.
- **Cross-example objective:** a chosen/rejected pairwise callback computes its scalar over the full logical batch and remains correct when the FSDP micro-batch is smaller than that logical batch.
- **Shape and masking:** returned log-probability lengths, target tokens, masks, and submitted gradient weights remain aligned for variable-length examples.
- **Forward-only safety:** the first half of the custom call does not accumulate gradients. This is particularly important after the SDK 0.25 routing change described in #145.
- **Failure sequencing:** an exception in the client callback produces a clear client error and does not leave the TuFT run in a state that prevents the next valid request.
- **Metrics:** callback metrics are preserved in the final `ForwardBackwardOutput` without changing server metric-reduction semantics.

## Acceptance criteria

- An unmodified TuFT server and supported Tinker SDK can run a client-defined log-probability loss through `forward_backward_custom` on HF and FSDP.
- No TuFT source patch, sentinel dispatch flag, or new SDK loss-name enum member is required.
- A cross-example composite DPO callback and response-normalized SFT callback pass an end-to-end FSDP smoke test.
- Documentation includes a minimal runnable example and clearly states that the custom callback runs client-side and incurs separate forward and backward requests.

This capability would unblock the draft Open Character Training example while also covering other research objectives that can be expressed as differentiable functions of target-token log probabilities.

## Related PR

- #147 adds the draft Open Character Training example that exercises this capability. The PR remains in draft until the custom-loss path is supported and passes its end-to-end FSDP validation.
